File: src/core/passes/outline.ts
// Shared skeleton for the outlining pass family.
//
// shared-terminal-tail, return-suffix-gadget, shared-straight-line-helper, and
// shared-call-tail all hunt for repeated op sequences and replace duplicates
// with a 2-cell transfer to one shared copy. They used to carry private copies
// of the same machinery: backward suffix collection keyed by op identity,
// greedy savings-ordered selection with range protection, collision-safe label
// allocation, and numeric-flow guards. This module owns those pieces; each
// pass keeps its own matcher (what counts as a shareable op / terminal) and
// its established label prefix and report identity.

import type { IrOp } from "../types.ts";
import { cellsPerOp } from "./helpers.ts";

export interface OutlineOccurrence {
  readonly key: string;
  readonly start: number;
  readonly end: number;
  readonly cells: number;
}

export interface OutlineCandidate {
  readonly key: string;
  readonly occurrences: OutlineOccurrence[];
  readonly cells: number;
}

export interface SelectedOutline {
  readonly label: string;
  readonly target: OutlineOccurrence;
  readonly replacements: OutlineOccurrence[];
}

export function targetKey(target: string | number): string {
  return typeof target === "number" ? `#${target}` : target;
}

/**
 * Outlining passes refuse to run on IR with raw numeric flow targets: moving
 * ops around would silently break hand-written cell addresses.
 */
export function hasNumericOutlineFlowTarget(ops: readonly IrOp[]): boolean {
  return ops.some((op) => {
    switch (op.kind) {
      case "jump":
      case "cjump":
      case "call":
      case "loop":
        return typeof op.target === "number";
      default:
        return false;
    }
  });
}

export function rangeIntersects(indexes: ReadonlySet<number>, start: number, end: number): boolean {
  for (let index = start; index <= end; index += 1) {
    if (indexes.has(index)) return true;
  }
  return false;
}

export function markRange(indexes: Set<number>, start: number, end: number): void {
  for (let index = start; index <= end; index += 1) indexes.add(index);
}

export interface LabelAllocator {
  /** Allocate the next collision-free `${prefix}${n}` label and reserve it. */
  next(): string;
}

/**
 * Label allocator seeded with every label already present in the program, so
 * repeated fixpoint iterations of an outlining pass can never emit a duplicate.
 */
export function createLabelAllocator(ops: readonly IrOp[], prefix: string): LabelAllocator {
  const existing = new Set(ops.flatMap((op) => op.kind === "label" ? [op.name] : []));
  let counter = 0;
  return {
    next: (): string => {
      let suffix = counter;
      while (existing.has(`${prefix}${suffix}`)) suffix += 1;
      counter += 1;
      const label = `${prefix}${suffix}`;
      existing.add(label);
      return label;
    },
  };
}

export interface SuffixCollectionConfig {
  /** Final op of a candidate suffix (e.g. return, or any terminal flow op). */
  readonly isTerminal: (op: IrOp) => boolean;
  /** Ops allowed inside the straight-line body before the terminal. */
  readonly isBodyOp: (op: IrOp) => boolean;
  /** Identity key for one op; sequences match iff their key lists match. */
  readonly opKey: (op: IrOp) => string;
}

/**
 * Collect every repeated-or-not straight-line suffix ending in a terminal op,
 * keyed by op identity. An occurrence is recorded once the sequence is longer
 * than the 2 cells its replacement jump would cost.
 */
export function collectSuffixCandidates(
  ops: readonly IrOp[],
  config: SuffixCollectionConfig,
): OutlineCandidate[] {
  const byKey = new Map<string, OutlineOccurrence[]>();

  for (let end = 0; end < ops.length; end += 1) {
    const final = ops[end]!;
    if (!config.isTerminal(final)) continue;

    const parts = [config.opKey(final)];
    let cells = cellsPerOp(final);
    for (let start = end - 1; start >= 0; start -= 1) {
      const op = ops[start]!;
      if (!config.isBodyOp(op)) break;

      parts.unshift(config.opKey(op));
      cells += cellsPerOp(op);
      if (cells <= 2) continue;

      const key = parts.join("\n");
      const occurrences = byKey.get(key) ?? [];
      occurrences.push({ key, start, end, cells });
      byKey.set(key, occurrences);
    }
  }

  return [...byKey.entries()]
    .map(([key, occurrences]) => ({ key, occurrences, cells: occurrences[0]!.cells }));
}

/**
 * Greedy shared-suffix selection: order candidates by total saved cells, keep
 * the first occurrence as the shared target, and replace the rest, protecting
 * every chosen range from later overlapping selections.
 */
export function selectSharedSuffixes(
  candidates: readonly OutlineCandidate[],
  labels: LabelAllocator,
  protectedIndexes: Set<number>,
): SelectedOutline[] {
  const selected: SelectedOutline[] = [];

  const ordered = [...candidates].sort((left, right) => {
    const leftSavings = (left.occurrences.length - 1) * (left.cells - 2);
    const rightSavings = (right.occurrences.length - 1) * (right.cells - 2);
    return rightSavings - leftSavings || right.cells - left.cells || left.key.localeCompare(right.key);
  });

  for (const candidate of ordered) {
    const available = candidate.occurrences.filter((occurrence) =>
      !rangeIntersects(protectedIndexes, occurrence.start, occurrence.end)
    );
    if (available.length < 2) continue;

    const [target, ...replacements] = available;
    if (target === undefined || replacements.length === 0) continue;

    const savedCells = replacements.reduce((sum, occurrence) => sum + occurrence.cells - 2, 0);
    if (savedCells <= 0) continue;

    selected.push({ label: labels.next(), target, replacements });

    for (const occurrence of [target, ...replacements]) {
      markRange(protectedIndexes, occurrence.start, occurrence.end);
    }
  }

  return selected;
}

File: native/src/core/passes/outline.cpp
#include "mkpro/core/passes/outline.hpp"

#include "mkpro/core/passes/helpers.hpp"

#include <algorithm>
#include <map>
#include <numeric>
#include <string>
#include <utility>
#include <variant>

namespace mkpro::core::passes {

LabelAllocator::LabelAllocator(const std::vector<IrOp>& ops, std::string prefix)
    : prefix_(std::move(prefix)) {
  for (const IrOp& op : ops) {
    if (op.kind == IrKind::Label)
      existing_.insert(op.name);
  }
}

std::string LabelAllocator::next() {
  int suffix = counter_;
  while (existing_.contains(prefix_ + std::to_string(suffix)))
    ++suffix;
  counter_ = suffix + 1;
  std::string label = prefix_ + std::to_string(suffix);
  existing_.insert(label);
  return label;
}

std::string target_key(const IrTarget& target) {
  if (const auto* numeric = std::get_if<int>(&target))
    return "#" + std::to_string(*numeric);
  return std::get<std::string>(target);
}

bool has_numeric_outline_flow_target(const std::vector<IrOp>& ops) {
  for (const IrOp& op : ops) {
    if ((op.kind == IrKind::Jump || op.kind == IrKind::CondJump || op.kind == IrKind::Call ||
         op.kind == IrKind::Loop) &&
        std::holds_alternative<int>(op.target)) {
      return true;
    }
  }
  return false;
}

bool range_intersects(const std::set<int>& indexes, int start, int end) {
  for (int index = start; index <= end; ++index) {
    if (indexes.contains(index))
      return true;
  }
  return false;
}

void mark_range(std::set<int>& indexes, int start, int end) {
  for (int index = start; index <= end; ++index)
    indexes.insert(index);
}

std::vector<OutlineCandidate> collect_suffix_candidates(const std::vector<IrOp>& ops,
                                                        const SuffixCollectionConfig& config) {
  std::map<std::string, std::vector<OutlineOccurrence>> by_key;

  for (int end = 0; end < static_cast<int>(ops.size()); ++end) {
    const IrOp& final = ops.at(static_cast<std::size_t>(end));
    if (!config.is_terminal(final))
      continue;

    std::vector<std::string> parts = {config.op_key(final)};
    int cells = cells_per_op(final);
    for (int start = end - 1; start >= 0; --start) {
      const IrOp& op = ops.at(static_cast<std::size_t>(start));
      if (!config.is_body_op(op))
        break;

      parts.insert(parts.begin(), config.op_key(op));
      cells += cells_per_op(op);
      if (cells <= 2)
        continue;

      std::string key;
      for (std::size_t index = 0; index < parts.size(); ++index) {
        if (index > 0)
          key += '\n';
        key += parts.at(index);
      }
      by_key[key].push_back(OutlineOccurrence{
          .key = key,
          .start = start,
          .end = end,
          .cells = cells,
      });
    }
  }

  std::vector<OutlineCandidate> result;
  result.reserve(by_key.size());
  for (auto& [key, occurrences] : by_key) {
    const int cells = occurrences.empty() ? 0 : occurrences.front().cells;
    result.push_back(OutlineCandidate{
        .key = std::move(key),
        .occurrences = std::move(occurrences),
        .cells = cells,
    });
  }
  return result;
}

std::vector<SelectedOutline> select_shared_suffixes(
    const std::vector<OutlineCandidate>& candidates, LabelAllocator& labels,
    std::set<int>& protected_indexes) {
  std::vector<SelectedOutline> selected;
  std::vector<OutlineCandidate> ordered = candidates;
  std::sort(ordered.begin(), ordered.end(), [](const OutlineCandidate& left,
                                               const OutlineCandidate& right) {
    const int left_savings =
        (static_cast<int>(left.occurrences.size()) - 1) * (left.cells - 2);
    const int right_savings =
        (static_cast<int>(right.occurrences.size()) - 1) * (right.cells - 2);
    if (right_savings != left_savings)
      return right_savings < left_savings;
    if (right.cells != left.cells)
      return right.cells < left.cells;
    return left.key < right.key;
  });

  for (const OutlineCandidate& candidate : ordered) {
    std::vector<OutlineOccurrence> available;
    for (const OutlineOccurrence& occurrence : candidate.occurrences) {
      if (!range_intersects(protected_indexes, occurrence.start, occurrence.end))
        available.push_back(occurrence);
    }
    if (available.size() < 2)
      continue;

    const OutlineOccurrence target = available.front();
    std::vector<OutlineOccurrence> replacements(available.begin() + 1, available.end());
    const int saved_cells = std::accumulate(
        replacements.begin(), replacements.end(), 0,
        [](int sum, const OutlineOccurrence& occurrence) { return sum + occurrence.cells - 2; });
    if (saved_cells <= 0)
      continue;

    selected.push_back(SelectedOutline{
        .label = labels.next(),
        .target = target,
        .replacements = replacements,
    });

    mark_range(protected_indexes, target.start, target.end);
    for (const OutlineOccurrence& replacement : replacements)
      mark_range(protected_indexes, replacement.start, replacement.end);
  }

  return selected;
}

} // namespace mkpro::core::passes
